### DISCUSSION TITLE
What is the best practice to clean the size of the logs?

### DISCUSSION THREAD

**Discussion by roysG** [emqx/emqx#8643] [Q&A] [answered]
I change the default settings and set that just one log file will be created and i set limit of 100 mb.

But as i checked, i see that the emqx.log is bigger than the limit i setted in the settings.

What will be the best practice to clean the log?

Thanks

**Comment by zhongwencool** [comment]
Hi, which version do you using?
please run those two command to see the log's detail config:
```
./bin/emqx versions

./bin/emqx eval "logger:get_handler_config()."

```
Normally, if you start with logrotate, he will recreate another `emqx.log.2` directly after reaching 100M (not exactly 100M, but not too much exceeds it).

**Reply by roysG** [reply]
Hi,
I am not using with the rotation option, cause is more easy for me to tracking logs in one file instead many files.
So i disable the option of the rotation.

I use in version: 
root@mqtt1:~# emqx versions
Installed versions:
* 4.3.5	permanent



emqx eval "logger:get_handler_config()."
[#{config =>
       #{burst_limit_enable => true,burst_limit_max_count => 500,
         burst_limit_window_time => 1000,drop_mode_qlen => 200,
         filesync_repeat_interval => no_repeat,flush_qlen => 1000,
         overload_kill_enable => false,overload_kill_mem_size => 3000000,
         overload_kill_qlen => 20000,overload_kill_restart_after => 5000,
         sync_mode_qlen => 10,type => standard_io},
   filter_default => stop,
   filters =>
       [{filter_non_ssl,{fun logger_filters:domain/2,{log,sub,[otp,ssl]}}}],
   formatter => {ssl_logger,#{}},
   id => ssl_handler,level => debug,module => logger_std_h},
 #{config =>
       #{burst_limit_enable => false,burst_limit_max_count => 20000,
         burst_limit_window_time => 1000,drop_mode_qlen => 3000,
         file => "/var/log/emqx/emqx.log",filesync_repeat_interval => 5000,
         flush_qlen => 8000,max_no_bytes => infinity,max_no_files => 1,
         overload_kill_enable => true,overload_kill_mem_size => 31457280,
         overload_kill_qlen => 20000,overload_kill_restart_after => 5000,
         sync_mode_qlen => 100,type => halt},
   filesync_repeat_interval => no_repeat,filter_default => log,
   filters => [{drop_progress_reports,{fun logger_filters:progress/2,stop}}],
   formatter =>
       {emqx_logger_textfmt,
           #{chars_limit => unlimited,single_line => true,
             template =>
                 [time," [",level,"] ",
                  {clientid,
                      [{peername,[clientid,"@",peername," "],[clientid," "]}],
                      [{peername,[peername," "],[]}]},
                  msg,"\n"]}},
   id => file,level => info,module => logger_disk_log_h}]

I am looking for a good way to remove/clean the emqx.log and create new one with the same permissions.

Just to mention, i also tried to remove the file and create new one and set the same permissions, but the size was not changed. 
Sound very strange...

**Comment by zmstone** [comment]
in 5.0 you can configure below two keys:
[log.file_handlers.$name.rotation.count](https://www.emqx.io/docs/en/v5.0/admin/cfg.html#log-rotation)
and [log.file_handlers.$name.max_size](https://www.emqx.io/docs/en/v5.0/admin/cfg.html#log-file-handler)

For 4.x it's `log.rotation.size` and `log.rotation.count`

**Reply by roysG** [reply]
Hi,
I disable the rotation, cause i prefer to see the log only in 1 file.
So i just needs to understand how to clean this file as best practice option.

I tried to delete it and set permissions for the emqx.log file. but somehow when the logs start again, it took the last bigger size.

**Reply by zmstone** [reply]
ah, sorry, I misread your question.
How did you clean it and how much bigger did it grow ?

**Reply by roysG** [reply]
root@mqtt1:/# ll /var/log/emqx/emqx.log
-rw-rw-r-- 1 emqx emqx 4424923930 Aug  4 21:14 /var/log/emqx/emqx.log

I need to clean this file to be 0kb. 

What is the right way to do it?

**Reply by roysG** [reply]
As you can see, the file size if very large:

ls -lh /var/log/emqx/emqx.log
-rw-rw-r-- 1 emqx emqx 4.2G Aug  4 21:20 /var/log/emqx/emqx.log

**Comment by zmstone** [comment, marked as answer]
This seems to be the logic:
* There is no size limit if rotation is turned off.
* If rotation count is configured to 1, the .1 log file size will be caped by max_bytes (log.rotation.size) and when it's reached, it goes back to start writing from be beginning of the file.

**Reply by roysG** [reply]
Ok, it looks that this solution works!
I will still follow it and update...

One more thing, Does 4.3.5 considering to be stable for production in clustering when have many connections?

**Reply by zmstone** [reply]
Please consider using the latest 4.4
4.3 has a bug in Erlang virtual machine which may affect clustering stability (for which we are working on a fix which is to be release in 4.3.17)

**Reply by roysG** [reply]
I have clustering of 4.3.5 in production, the upgrade to 4.4 considring to be risky or easy update?
Can you send me the instruction how to do it? thanks.

**Reply by zmstone** [reply]
Here are the tested steps.
https://www.emqx.io/docs/en/v4.4/changes/upgrade-4.4.html

**Reply by roysG** [reply]
Thanks!

**Reply by roysG** [reply]
Do you have also instruction upgrading to v5?

**Reply by zmstone** [reply]
To v5, it will have to be a new cluster.
It’s more of a “migration” than “upgrade”
https://www.emqx.io/docs/en/v5.0/deploy/upgrade-from-v4.html
